### PR TITLE
Add combobox (search bar w/ autocomplete  + selection menu) for tag s…

### PR DESCRIPTION
**Pull request by ryan-draper**
…election.

Tag selection is now a 2-step process: (1) selecting tags from a searchable select menu and (2) ranking the selected tags.

Deployed: https://donations-step-2020-new.appspot.com/

I have also written proper handling of incorrect tag entries (entering the wrong number of tags, entering the same tag twice, etc).

### CODE REVIEW

**Comment by ryan-draper** [comment]
Also, here are the general UI layouts I was considering for the homepage. I ended up going with (A), but I'd be happy to change it if y'all prefer another one.

(A)
Screenshot 2020-07-18 at 2 23 06 AM

(B)
Screenshot 2020-07-18 at 2 23 25 AM

(C)
Screenshot 2020-07-18 at 2 23 43 AM

(D)
Screenshot 2020-07-18 at 2 24 16 AM

**Comment by ryan-draper** [comment]
The only "bug" I am aware of now is that the "2. Click and drag to order the causes" is also a sortable element; therefore, if a tag is dragged above it, then it is considered one of the three tags. The algorithm still works, but it ignores whatever actual tag is the last in the ranking. I'll address this issue in my next PR.

File: Donation-platform/src/main/webapp/personalizationScript.js [hunk 1]
```diff
@@ -107,9 +135,203 @@ function displayTags(tags) {
...
-    loadPersonalizedCharities(tagOrder);
+    var tags = [];
+    for(i = 1; i < tagOrder.length; i++) {
+      tags.push(tagOrder[i]);
+    }
+
+    // If there are 3 tags in the rank, run loadPersonalizedCharities.
+    if(tags.length == 3) {
+      loadPersonalizedCharities(tags);
+    // Otherwise, alert the user to select 3 causes before re-submitting.
```
Comment (charu0291) [inline Donation-platform/src/main/webapp/personalizationScript.js:150]: we should not force user to select atleast 3. It should be maximum of 3. Can we have "2 nd part " as dynamic to show only 1/2/3 ? if not then we should choose other 2 randomly.

Comment (ryan-draper) [reply Donation-platform/src/main/webapp/personalizationScript.js]: Done.

**Comment by charu0291** [comment]
Please wait for others to also review the PR's. Feel free to ping them personally if they don't review it in 24hrs.

File: Donation-platform/src/main/webapp/personalizationScript.js [hunk 2]
```diff
@@ -91,6 +95,34 @@ function updatePersonalizedCardsOnPage(charities) {
...
+function showRankInstructions(tag) {
+    const tagRank = document.getElementById('sortable');
+    tagRank.innerHTML += '<label><b>2. Click and drag to order the causes:</b></label>';
+}
+
+/* Creates and displays the tag element in the rank given the tag's name */
+function placeSelectedTagInRank(tag) {
+    const tagRank = document.getElementById('sortable');
+
+    let tagElements = new Map([
```
Comment (tonioshikanlu) [inline Donation-platform/src/main/webapp/personalizationScript.js:108]: Could this be dynamically created? I can imagine if we have hundreds of tags we would not be able to hardcode every single one.

File: Donation-platform/src/main/webapp/personalizationScript.js [hunk 3]
```diff
@@ -107,9 +139,211 @@ function displayTags(tags) {
...
+    // 3 causes before re-submitting (although 1 or 2 tags is also allowed, 
+    // encourage 3 so the user gets the most comprehensive results).
+    } else if(tags.length == 0) {
+      alert("Please select 3 causes to get personalized charities.");
+    // Otherwise, if there more than 3 tags in the rank, alert the user to select
+    // no more than 3 causes before re-submitting (NOTE: the function that places 
+    // tags into the ranking already prevents more than 3 tags from being added; 
+    // this is a fail-safe).
+    } else if(tags.length > 3) {
+      alert("Please select no more than 3 tags to get personalized charities");
```
Comment (tonioshikanlu) [inline Donation-platform/src/main/webapp/personalizationScript.js:164]: If you want to look into stylizing your alerts i can recommend sweetalert2 if you're interested. That's what I used in the addCharity.html page.

File: Donation-platform/src/main/webapp/personalizationScript.js (hunk 2 above)
Comment (ryan-draper) [reply Donation-platform/src/main/webapp/personalizationScript.js:109]: The problem is that right now we don't store any sort of tag image/icon in the Tag class (correct me if I'm wrong). So regardless I'll need to manually create the icons associated with each tag. Once we coordinate tag images/icons and store them in the db I'll change this to be dynamic.

**Comment by ryan-draper** [comment]
> The only "bug" I am aware of now is that the "2. Click and drag to order the causes" is also a sortable element; therefore, if a tag is dragged above it, then it is considered one of the three tags. The algorithm still works, but it ignores whatever actual tag is the last in the ranking. I'll address this issue in my next PR.

Fixed.

File: Donation-platform/src/main/webapp/personalizationScript.js (hunk 2 above)
Comment (tonioshikanlu) [reply Donation-platform/src/main/webapp/personalizationScript.js:109]: We have an image source in the tag class. It'll be in my next merge.

**Comment by ryan-draper** [comment]
Screenshot 2020-07-21 at 3 16 32 PM

File: Donation-platform/src/main/webapp/index.html [hunk 4]
```diff
@@ -9,6 +9,7 @@
     <link rel="stylesheet" href="personalizedStyle.css">
     <script src="userAuth.js"></script>
     <meta name="viewport" content="width=device-width, initial-scale=1">
+    <script src="https://unpkg.com/sweetalert/dist/sweetalert.min.js"></script>
```
Comment (curves4d) [inline Donation-platform/src/main/webapp/index.html:12]: include all third party scripts before putting any of your own scripts

Comment (curves4d) [reply Donation-platform/src/main/webapp/index.html:12]: also for css files

File: Donation-platform/src/main/webapp/personalizationScript.js [hunk 5]
```diff
@@ -91,6 +82,47 @@ function updatePersonalizedCardsOnPage(charities) {
     cards.innerHTML += toAdd;
 }
 
+/* Converts the list of sortable-compitable tag values (which require underscores) 
+   into a list of the top 3 tags with their proper names.
+
+   Example: tagOrder = ["li_health", "li_hunger", "li_education", "li_children",
```
Comment (curves4d) [inline Donation-platform/src/main/webapp/personalizationScript.js:88]: give an input output example

File: Donation-platform/src/main/webapp/personalizationScript.js (hunk 2 above)
Comment (ryan-draper) [reply Donation-platform/src/main/webapp/personalizationScript.js:108]: Ok, I am taking note to update this function once we do your merge.

File: Donation-platform/src/main/webapp/index.html (hunk 4 above)
Comment (ryan-draper) [reply Donation-platform/src/main/webapp/index.html]: Done.

File: Donation-platform/src/main/webapp/personalizationScript.js (hunk 5 above)
Comment (ryan-draper) [reply Donation-platform/src/main/webapp/personalizationScript.js]: I made it more clear that 'tagOrder' and 'tags' are respectively the input and output by replacing their names with input/output in the comment.

**Pull request merged**